File: Data Structures & Algorithms/combination-target-sum-ii/submission-5.java
class Solution {
    Set<List<Integer>> ans;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        ans = new HashSet<>();
        Arrays.sort(candidates);
        helper(candidates, target, 0, new ArrayList<>());
        return new ArrayList<>(ans);
    }
    public void helper(int[] candidates, int target, int i, List<Integer> curr) {
        if(target==0) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        if(target<0 || i==candidates.length) return;
        curr.add(candidates[i]);
        helper(candidates, target-candidates[i], i+1, curr);
        curr.removeLast();
        helper(candidates, target, i+1, curr);
    }
}

File: Data Structures & Algorithms/combination-sum-iv/submission-5.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        dp = [-1]*(target+1)

        def traverse(curr):
            if curr>target:
                return 0
            if dp[curr]!=-1:
                return dp[curr]
            if curr==target:
                return 1
            ans = 0
            for j in range(n):
                ans += traverse(curr+nums[j])
            dp[curr] = ans
            return ans

        return traverse(0)